src/tsdb.c

```diff
@@ -224,75 +224,85 @@ int SeriesAddSample(Series *series, api_timestamp_t timestamp, double value) {
     return TSDB_OK;
 }
 
-SeriesIterator SeriesQuery(Series *series, api_timestamp_t minTimestamp, api_timestamp_t maxTimestamp) {
-    SeriesIterator iter;
+int SeriesQuery(Series *series, SeriesIterator *iter, 
+                            api_timestamp_t minTimestamp, api_timestamp_t maxTimestamp) {
     timestamp_t rax_key;
-    iter.series = series;
+    iter->series = series;
+    ChunkFuncs *funcs = iter->series->funcs;
+    
     // get the rightmost chunk whose base timestamp is smaller or equal to minTimestamp
     seriesEncodeTimestamp(&rax_key, minTimestamp);
-    iter.dictIter = RedisModule_DictIteratorStartC(series->chunks, "<=", &rax_key, sizeof(rax_key));
-
-    // if no such chunk exists, we will start the search from the first chunk
-    if (!RedisModule_DictNextC(iter.dictIter, NULL, (void*)&iter.currentChunk))
-    {
-        RedisModule_DictIteratorReseekC(iter.dictIter, "^", NULL, 0);
-        RedisModule_DictNextC(iter.dictIter, NULL, (void*)&iter.currentChunk);
+    iter->dictIter = RedisModule_DictIteratorStartC(series->chunks, "<=", &rax_key, sizeof(rax_key));
+    RedisModule_DictNextC(iter->dictIter, NULL, (void*)&iter->currentChunk);
+    
+    // iterate to the first relevant chunk
+    void *dictResult = (void *)TRUE;
+    while (dictResult) {
+        if (funcs->GetLastTimestamp(iter->currentChunk) < minTimestamp) {
+            dictResult = RedisModule_DictNextC(iter->dictIter, NULL, (void*)&iter->currentChunk);
+        } else {
+            iter->chunkIterator = funcs->NewChunkIterator(iter->currentChunk);
+            iter->minTimestamp = minTimestamp;
+            iter->maxTimestamp = maxTimestamp;
+            return REDISMODULE_OK;
+        }
     }
-    iter.chunkIteratorInitialized = FALSE;
-    iter.chunkIterator = NULL;
-    iter.minTimestamp = minTimestamp;
-    iter.maxTimestamp = maxTimestamp;
-    return iter;
+    SeriesIteratorClose(iter);
+    return REDISMODULE_ERR;
 }
 
-void SeriesIteratorClose(SeriesIterator *iterator) {
-    iterator->series->funcs->FreeChunkIterator(iterator->chunkIterator);
-    RedisModule_DictIteratorStop(iterator->dictIter);
-}
+ChunkResult SeriesIteratorGetFirst(SeriesIterator *iterator, Sample *sample) {
+    ChunkFuncs *funcs = iterator->series->funcs;
```

Comment:
@ashtul IMO this is not the correct way to work with iterators.
This function does 2 things:
1. jumps to the start of the stream
2. returns the first one.

This should be a function that jumps to the start and then use the GetNext to actually retrieve the sample.
Another reason for that is that you can use `GetFirst` after you used other iterator function and expect that the original order will be saved, every time you call this function you reset the iterator "cursor".

Reply:
@danni-m 
`SeriesQuery` jumps to the start of the stream/chunk while `SeriesIteratorGetFirst` returns from that chunk the first sample
We had a discussion about this b/c we thought to incorporate `getFirst` into `SeriesQuery`.
For then last point, getFirst does not reset the iterator but will not move to the next chunk as will getNext.